Add config flow for workday integration

javicalle opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds config_flow configuration to workday integration.

It is my first implementation of a config flow and the review is appreciated.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
This started as an academic exercise to understand the 'config_flow' mechanisms and how to implement it.
He believed that this integration was going to be an easy case, but it has been more 'complete' than he initially expected.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

### Other changes:
* constant values exported to `cons`
* `unique_id` (added to class constructor)

### TODO:
- [x] WorkdayConfigFlow.CONNECTION_CLASS value
- [x] better date validation
- [x] general code review (comments, logs, ...)
- [x] config_flow tests
- [ ] update docs

### Implementations details
* Config flow offers a basic configuration flow and the possibility to configure 'advanced' options in a second flow step
  * _basic_ step configure `state` and `province`
  * _advanced_ step allows configure `workdays`, `excludes` and `offset`
* Options flow allow configure 'extra holidays' and 'remove holidays'
* When imported from YAML, configuration options are initially stored in `ConfigEntry`. After, in `async_setup_entry` the `_async_import_options_from_data_if_missing` method will pop attributes from `ConfigEntry` and stored as `options`

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

javicalle committed:
Add config_flows to workday sensor

javicalle committed:
fix execute permission

javicalle committed:
adapt tests to the config_flow implementation

javicalle committed:
fix execute

javicalle committed:
added tests

javicalle committed:
fix validations

javicalle commented:
Not stale.
Waiting for review.

javicalle commented:
Not stale.
Waiting for review.

javicalle committed:
Fix pylint checks

javicalle commented:
Not stale.
Rebase done.

gjohansson-ST commented:
Would like to test this one but it seems you have a mypy error in the checks

javicalle committed:
fix mypy tests

javicalle commented:
Fixed test failures introduced with the last rebase.
It may be that some of the implemented workarounds can now be addressed in another way. I haven't reviewed workflow implementations for quite some time.

javicalle committed:
fix workday tests

javicalle commented:
Not stale.

milanmeu commented on homeassistant/components/workday/__init__.py:
```diff
@@ -1 +1,87 @@
+    # they are missing from the options
+    _async_import_options_from_data_if_missing(hass, entry)
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
```
```suggestion
    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```

milanmeu commented on homeassistant/components/workday/__init__.py:
```diff
@@ -1 +1,87 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
```
```suggestion
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```

milanmeu commented on homeassistant/components/workday/config_flow.py:
```diff
@@ -0,0 +1,371 @@
+    ERR_NO_COUNTRY,
+    ERR_NO_SUBCOUNTRY,
+)
+from .const import DOMAIN  # pylint:disable=unused-import
```
```suggestion
from .const import DOMAIN
```

milanmeu commented on homeassistant/components/workday/config_flow.py:
```diff
@@ -0,0 +1,371 @@
+    """Handle a config flow for workday."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
```
This moved to the manifest
```suggestion
```

milanmeu reviewed (commented):
I don't have the necessary experience with yaml to config flow so I can't review more.
There are still some untested lines in `config_flow.py`, 100% coverage is [required](https://developers.home-assistant.io/docs/development_checklist/).

javicalle committed:
fix config_flow 100% coverage

javicalle commented:
Finally I get some time to rebase and review. 

I have workarounded `mypy` with:
```
    for date in remove_holidays:  # type: ignore
```
I have no idea how this can be addressed. Any help will be appreciated.

@milanmeu sorry for the delay, but if you still want to review the implementation, you are welcome.

gjohansson-ST commented:
> Finally I get some time to rebase and review.
> 
> I have workarounded `mypy` with:
> 
> ```
>     for date in remove_holidays:  # type: ignore
> ```
> 
> I have no idea how this can be addressed. Any help will be appreciated.
>
I guess perhaps you need to set a typehint on this line
```
remove_holidays: list[datetime.date] = config_entry.options.get(CONF_REMOVE_HOLIDAYS)
```

javicalle commented:
@gjohansson-ST thanks for your reply
I have tried with:
```
    remove_holidays: list[str] = config_entry.options.get(CONF_REMOVE_HOLIDAYS)
```
but then I get:
```
$ mypy homeassistant/components/workday/*.pyhomeassistant/components/workday/binary_sensor.py:92: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "List[str]")  [assignment]
Found 1 error in 1 file (checked 4 source files)
```

I will look for other typehints to see if I find the solution.

Thanks.

gjohansson-ST commented:
> @gjohansson-ST thanks for your reply
> 
> I have tried with:
> 
> ```
> 
>     remove_holidays: list[str] = config_entry.options.get(CONF_REMOVE_HOLIDAYS)
> 
> ```
> 
> but then I get:
> 
> ```
> 
> $ mypy homeassistant/components/workday/*.pyhomeassistant/components/workday/binary_sensor.py:92: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "List[str]")  [assignment]
> 
> Found 1 error in 1 file (checked 4 source files)
> 
> ```
> 
> 
> 
> I will look for other typehints to see if I find the solution.
> 
> 
> 
> Thanks.

Optional[List[datetime.date]] I would guess then. Didn't think about that it's an option.

javicalle committed:
remove  '# type: ignore' workaround

milanmeu commented on homeassistant/components/workday/binary_sensor.py:
```diff
@@ -208,3 +218,8 @@ async def async_update(self):
             self._state = False
+
+    @property
+    def unique_id(self) -> str:
```
Use the `self._attr_unique_id` instance attribute.
https://developers.home-assistant.io/docs/core/entity/#entity-class-or-instance-attributes

javicalle commented:
You are right!!! I was not convinced of the `datetime.date` value but it works:
```
remove_holidays: vol.Optional[List[datetime.date]] = config_entry.options.get(CONF_REMOVE_HOLIDAYS)
```
However, I still believe that the correct value is `str`, the list can contain the dates or the name of the holidays:
````
                # is this formatted as a date?
                if dt.parse_date(date):
                    # remove holiday by date
                    removed = obj_holidays.pop(date)
                else:
                    # remove holiday by name
                    _LOGGER.debug("Treating '%s' as named holiday", date)
                    removed = obj_holidays.pop_named(date)
```

So in the end the code I have left it as follows:
```
     remove_holidays: vol.Optional [list [str]] = config_entry.options.get (
         CONF_REMOVE_HOLIDAYS
     )
```

Thank you very much for your help. I greatly appreciate it.

milanmeu commented on homeassistant/components/workday/__init__.py:
```diff
@@ -1 +1,76 @@
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the workday component."""
+    hass.data.setdefault(DOMAIN, {})
```
I think this is not used.
```suggestion
```

javicalle committed:
Use the self._attr_unique_id instance attribute

milanmeu commented on homeassistant/components/workday/__init__.py:
```diff
@@ -1 +1,76 @@
+    """Set up the workday component."""
+    hass.data.setdefault(DOMAIN, {})
+
+    # Import configuration from sensor platform
+    config_platform = config_per_platform(config, "binary_sensor")
+    for p_type, p_config in config_platform:
+        if p_type != DOMAIN:
+            continue
```
You can call `hass.config_entries.flow.async_init` in `async_setup_platform` (in binary_sensor.py) so you don't need to search the config of every HA installations even the ones without workday.

Add this to binary_sensor.py:
```python3
async def async_setup_platform(hass, config, add_entities, discovery_info=None) -> None:
    await hass.config_entries.flow.async_init(
                DOMAIN,
                context={"source": config_entries.SOURCE_IMPORT},
                data=config,
            )
```

javicalle commented:
I really appreciate all the comments, but I am more concerned with the usability and the management of options and settings.
Any opinion or comment in this regard is appreciated.

milanmeu commented on homeassistant/components/workday/__init__.py:
```diff
@@ -1 +1,76 @@
+from homeassistant.helpers import config_per_platform
+
+from .const import CONF_ADD_HOLIDAYS, CONF_REMOVE_HOLIDAYS, DOMAIN
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
```
```suggestion
```

javicalle committed:
replace async_setup with async_setup_platform

javicalle commented:
Some test error with python 3.8:
```
Error: homeassistant/components/workday/binary_sensor.py:92:21: E1136: Value 'vol.Optional' is unsubscriptable (unsubscriptable-object)
```

Any advice here?

milanmeu commented:
> Some test error with python 3.8: 
> ```
> Error: homeassistant/components/workday/binary_sensor.py:92:21: E1136: Value 'vol.Optional' is unsubscriptable (unsubscriptable-object)
> ```

You use Optional from voluptuous, you should use Optional from typing.
However, we prefer `list[str] | None` instead of `Optional[list[str]]` because Optional is not allowed by pyupgrade.
Note that this union syntax requires Python >=3.10 or `from __future__ import annotations` in 3.8 and 3.9.

milanmeu commented on homeassistant/components/workday/config_flow.py:
```diff
@@ -0,0 +1,369 @@
+                return await self.async_step_add_holidays()
+            if action == ACTION_REMOVE_HOLIDAYS:
+                return await self.async_step_remove_holidays()
+            # errors[OPTIONS_ACTION] = "unsupported_action"
```
PR shouldn't have commented out code.

javicalle committed:
fix pylint 3.8 check

javicalle commented:
Finally get fixed with a combination of code checks (validating None value, fixing too many nested blocks, ...).
`list[str] | None` working locally, let's see the tests.

gjohansson-ST commented:
> I really appreciate all the comments, but I am more concerned with the usability and the management of options and settings.
> Any opinion or comment in this regard is appreciated.

What exactly are you referring too? You mean how the config_flow works with data and options?

javicalle commented:
> > I really appreciate all the comments, but I am more concerned with the usability and the management of options and settings.
> > Any opinion or comment in this regard is appreciated.
> 
> What exactly are you referring too? You mean how the config_flow works with data and options?

Well, all the UI part.
In the moment that I implemented this config flow I couldn't find any similar case, and I try my best to get a working implementation.
Not sure if it is the best approach or if there's new API methods that facilitates (for example) list values management (this was a pain in the ass).
I don't have reviewed other config flows for months and I am not aware of code changes.

gjohansson-ST commented:
> Well, all the UI part.
> In the moment that I implemented this config flow I couldn't find any similar case, and I try my best to get a working implementation.
> Not sure if it is the best approach or if there's new API methods that facilitates (for example) list values management (this was a pain in the ass).
> I don't have reviewed other config flows for months and I am not aware of code changes.

I looked at this myself for this integration but did not come to a good way to deal with lists hence I never got too it but now you are doing it instead so perfect.
The only example I have seen so far is actually to have an additional checkbox and then trigger the form again for every time it checked and hence build the list from those. But it's a quite poor user experience if you have a long list and you need a new form for every new date to add or remove.
Let me see if I can find an integration with a list to see how it's been done by someone else.

frenck commented:
For this PR to be considered, this architecture issue needs to be resolved: <https://github.com/home-assistant/architecture/discussions/569>

In the current state of affairs, we don't allow this integration to use a configuration flow.

javicalle committed:
fix imported entity unique_id

javicalle commented:
I have read the (short) arquitecture discussion https://github.com/home-assistant/architecture/discussions/569
Workday integration does not connect devices or services, therefore it is not allowed to be implemented as a config_flow. It's a pitty, but I should have been aware of that before I started implementing. 

In any case, @frenck thank you very much for taking the time to let me know, I appreciate it. 
I'm sorry if I can't contribute to the arquitecture discussion, but I think it's a bit too big for me.

And thanks to the other reviewers for their comments, their help and for being there.

I will close the issue pending the resolution of the discussion.

gjohansson-ST commented:
Too bad. I was looking forward to this one. But I guess it can re-open once the arch decision has been made.
In general I don't see why a config flow would not be allowed for any integration regardless service, device or self-evaluating.

frenck commented:
@gjohansson-ST That is off-topic for this PR. If you like to discuss general Home Assistant architecture, we have a dedicated place for that. Thanks 👍
